Created and implemented the PropertyDictionary class

njlawton opened the pull request:
This class contains methods for checking if strings contain vaild CSS propertys as well as methods for testing valid inputs of style values into CSS propertys.
A test class was also created to validate the methods created

Commit:
Created and implemented the PropertyDictionary class
This class contains methods for checking if strings contain vaild CSS propertys as well as methods for testing valid inputs of style values into CSS propertys.
A test class was also created to validate the methods created

wilddamon commented on src/property-dictionary.js:
```diff
@@ -0,0 +1,95 @@
+  }
+  var instance;
+
+  PropertyDictionary.prototype.isProperty = function(property) {
```
What is the purpose of this function? Is it to check that the property is a supported property? If so, please change the name to "isSupportedProperty".

wilddamon commented on src/property-dictionary.js:
```diff
@@ -0,0 +1,95 @@
+
+  PropertyDictionary.prototype._lengthValueHasPercentage = function(lengthValue) {
+    if (!(lengthValue instanceof LengthValue)) {
+      return false;
```
Perhaps this should throw an exception to make it clear that the caller is doing something silly.

wilddamon commented on src/property-dictionary.js:
```diff
@@ -0,0 +1,95 @@
+      return (lengthValue.percent != null);
+    }
+
+    return (lengthValue.type == 'percent');
```
Can you use the enum value instead of writing the string directly here?

wilddamon commented on src/property-dictionary.js:
```diff
@@ -0,0 +1,95 @@
+  };
+
+  PropertyDictionary.prototype.isValidInput = function(property, styleValue) {
+    if (this.isProperty(property) == false) {
```
if (!this.isProperty(property)) {

Commit:
Changed _lengthValueHasPercentage method to throw an error for non LengthValue inputs
This change did require making adjustments in the isValidInput method to ensure that errors were not thrown incorrectly

Commit:
Made small changes to the PropertyDictionary class due to pull request feedback

wilddamon commented on src/property-dictionary.js:
```diff
@@ -0,0 +1,95 @@
+      if (!(this.propertyKeywords.hasOwnProperty(property))) {
+        return false;
+      }
+      return (this.propertyKeywords[property].indexOf(styleValue.keywordValue) > -1);
```
how about making a helper method for this check

```
PropertyDictionary.prototype.isValidKeyword = function(property, str) {
  return this.propertyKeywords[property].indexOf(str) >= 0;
}
```

then this function becomes

```
if (styleValue instanceof KeywordValue) {
  return this.isProperty(property) ? this.isValidKeyword(styleValue.keywordValue) : false;
}
```

wilddamon commented on src/property-dictionary.js:
```diff
@@ -0,0 +1,95 @@
+
+  var getInstance = function() {
+    if (!instance) {
+     instance = createInstance();
```
The constructor for PropertyDictionary is quite simple, so it should be the same to delete createInstance, and make this function read

```
var getInstance = function() {
  if (!instance) {
    instance = new PropertyDictionary();
  }
  return instance;
}
```

wilddamon commented on src/property-dictionary.js:
```diff
@@ -0,0 +1,95 @@
+    return instance;
+  };
+
+  shared.getInstance = getInstance;
```
This function name is too general - what if we want some other singleton class?

How about something like "getPropertyDictionaryInstance()", or as is common in blink, "propertyDictionary()"

wilddamon commented on test/js/property-dictionary.js:
```diff
@@ -0,0 +1,54 @@
+  });
+
+  test('Testing the isValidInput method with a non Length value property' , function() {
+    assert.isTrue(propertyDictionary.isValidInput('pitch-range', new NumberValue(5)));
```
Just add this line to the above test - it's essentially testing the same thing as

`The isValidInput method should return true if the property can accept the style value entered`

wilddamon commented on test/js/property-dictionary.js:
```diff
@@ -0,0 +1,54 @@
+    assert.isTrue(propertyDictionary.isValidInput('pitch-range', new NumberValue(5)));
+  });
+
+  test('The isValidInput method should return false if the property cant accept the style value entered' , function() {
```
s/cant/can't

Commit:
Added helper method _isValidKeyword to improve readablity of PropertyDictionary class
Also changed singleton method name to the more specific propertyDictionary method

wilddamon commented on src/property-dictionary.js:
```diff
@@ -0,0 +1,94 @@
+(function(shared, scope, testing) {
+
+  function PropertyDictionary() {
+    this.propertyMap = {
```
Can we change the name of this to be "validProperties" for clarity? It's not very helpful to have a PropertyDictionary that contains a propertyMap...

wilddamon replied:
These three maps should be "private", you can indicate this by prepending their names with an underscore;

this._validProperties
this._validKeywords
this._allowsPercentage

wilddamon commented on src/property-dictionary.js:
```diff
@@ -0,0 +1,94 @@
+      if (!(styleValue instanceof styleValueType)) {
+        continue;
+      }
+      if (!(styleValue instanceof LengthValue)) {
```
The logic here is confusing. I know you're trying to reduce nesting, but in this case all the negation makes it more confusing... how about something like

```
...
  for (var i = 0; i < this.propertyMap[property].length; i++) {
    var styleValueType = this.propertyMap[property][i];
    if (styleValue instanceOf styleValueType) {
      if (styleValueType != LengthValue) {
        return true;
      }
      return (!this._lengthValueHasPercentage(styleValue) || this.allowsPercentage[property]);
    }
  }
  return false;
};
```

FWIW, you can also return false if you find a percentage LengthValue when allowsPercentage returns false.

Also you don't need to check hasOwnProperty for allowsPercentage, you can just do allowsPercentage[property], which returns `undefined`, which is "falsy" (equivalent to `false`).

wilddamon commented on src/property-dictionary.js:
```diff
@@ -0,0 +1,94 @@
+
+  PropertyDictionary.prototype._lengthValueHasPercentage = function(lengthValue) {
+    if (!(lengthValue instanceof LengthValue)) {
+      throw new TypeError('The input to this method must be an object of type LengthValue');
```
It's better to say

'The input to _lengthValueHasPercentage must be an object of type LengthValue'

for clarity, since you know exactly where to go looking for you error in that case.

Commit:
Changed names of dicitionarys stored in the PropertyDictionary class
Changed name of propertyMap to _validProperties and also changed other maps to have a \'_\' character before there name indicating them as private

Commit:
Improved readability of the isValidInput method in the PropertyDictionary class

njlawton commented:
Finished those changes are there any other major problems? otherwise I will work on the next pull request and merge before submitting pull request for inspection.

Commit:
Merge branch 'master' into property-dictionary

wilddamon commented on src/property-dictionary.js:
```diff
@@ -0,0 +1,91 @@
+
+  PropertyDictionary.prototype._lengthValueHasPercentage = function(lengthValue) {
+    if (!(lengthValue instanceof LengthValue)) {
+      throw new TypeError('The input to _lengthValueHasPercentage must be an object of type LengthValue');
```
Line too long, break after (

wilddamon commented on src/property-dictionary.js:
```diff
@@ -0,0 +1,91 @@
+        if (styleValueType != LengthValue) {
+          return true;
+        }
+        return (!this._lengthValueHasPercentage(styleValue) || this._allowsPercentage.hasOwnProperty(property));
```
Line too long, break after ||

wilddamon commented on src/property-dictionary.js:
```diff
@@ -0,0 +1,91 @@
+  };
+
+  shared.propertyDictionary = propertyDictionary;
+  if (TYPED_OM_TESTING)
```
needs braces {

wilddamon commented:
Three really minor comments, otherwise LGTM

Commit:
Fixed small formatting issues found during pull request

njlawton commented:
Some lines i.e. defining methods are still slightly too long, also the line with the error message I broke is still slightly too long (none more than 90 characters) is this ok?

Commit:
Shorted some lines in the PropertyDictionary class

Commit:
Fixed minor formatting issue

wilddamon commented:
LGTM
